compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -373,11 +413,108 @@ impl<'tcx> SizeSkeleton<'tcx> {
             (SizeSkeleton::Pointer { tail: a, .. }, SizeSkeleton::Pointer { tail: b, .. }) => {
                 a == b
             }
+            // constants are always pre-normalized into a canonical form so this
+            // only needs to check if their pointers are identical.
+            (SizeSkeleton::Generic(a), SizeSkeleton::Generic(b)) => a == b,
             _ => false,
         }
     }
 }
 
+/// To compare constants, we only use this to multiply them, which ensures
+/// that each multiplied term is in sorted order.
+/// i.e. b * a always gets normalized to a * b, in order of the Ord impl
+/// of `Const`.
+/// It will return `None` if multiplication of constants overflows.
+fn mul_sorted_consts<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+    a: ty::Const<'tcx>,
+    b: ty::Const<'tcx>,
+) -> Option<ty::Const<'tcx>> {
+    use crate::mir::BinOp::Mul;
+    use ty::ConstKind::Expr;
+    use ty::Expr::Binop;
+
+    // fast-path
+    match (a.kind(), b.kind()) {
+        (Expr(Binop(Mul, ..)), _) | (_, Expr(Binop(Mul, ..))) => {}
+        _ => {
+            let (l, r) = if a < b { (a, b) } else { (b, a) };
+            return Some(tcx.mk_const(Expr(Binop(Mul, l, r)), tcx.types.usize));
+        }
+    }
+
+    let mut buf: SmallVec<[ty::Const<'tcx>; 4]> = SmallVec::new();
+    let mut a = sort_mul(tcx, param_env, a)?;
+    let mut b = sort_mul(tcx, param_env, b)?;
+    while let (Expr(Binop(Mul, al, ar)), Expr(Binop(Mul, bl, br))) = (a.kind(), b.kind()) {
+        if al < bl {
+            buf.push(al);
+            a = ar;
+        } else {
+            buf.push(bl);
+            b = br;
+        }
+    }
+
+    while let Expr(Binop(Mul, al, ar)) = a.kind() {
+        if al < b {
+            buf.push(al);
+        } else {
+            buf.push(std::mem::replace(&mut b, al));
+        }
+        a = ar;
+    }
+    while let Expr(Binop(Mul, bl, br)) = b.kind() {
+        if bl < a {
+            buf.push(bl);
+        } else {
+            buf.push(std::mem::replace(&mut a, bl));
+        }
+        b = br;
+    }
+
+    if a < b {
+        buf.push(a);
+        buf.push(b);
+    } else {
+        buf.push(b);
+        buf.push(a);
+    }
+
+    // buf is now sorted minimum -> maximum, with evaluatable at the end.
+    // compress all multiplied constants into a single value.
+    let mut const_mul: u64 = 1;
+    while let Some(v) = buf.last().and_then(|c| c.try_eval_usize(tcx, param_env)) {
+        const_mul = const_mul.checked_mul(v)?;
+        buf.pop();
+    }
+    if const_mul != 1 {
+        buf.push(ty::Const::from_usize(tcx, const_mul));
+    }
+    // create a single tree from the buffer
+    buf.into_iter().rev().reduce(|acc, n| tcx.mk_const(Expr(Binop(Mul, n, acc)), n.ty()))
+}
+
+fn sort_mul<'tcx>(
```

Comment:
I think having this function is not great. You should be able to only have one recursive function, that makes it easier to understand the code than having two separate recursive functions which call each other.